core/utils/xml.js

```diff
@@ -76,5 +84,29 @@ Blockly.utils.xml.textToDomDocument = function(text) {
  */
 Blockly.utils.xml.domToText = function(dom) {
   var oSerializer = new XMLSerializer();
-  return oSerializer.serializeToString(dom);
+  return Blockly.utils.xml.sanitizeText(oSerializer.serializeToString(dom));
+};
+
+/**
+ * Sanitizes a text string so that it can be serialized to XML. Specifically,
+ * this replaces control characters with their NCR (numerical character
+ * reference).
+ * @param {string} text The text to sanitize.
+ * @return {string} The sanitized text.
+ */
+Blockly.utils.xml.sanitizeText = function(text) {
+  if (!text.match(Blockly.utils.xml.INVALID_CONTROL_CHARS)) {
+    return text;
+  }
+  var match;
+  var newText = '';
+  while ((match = text.match(Blockly.utils.xml.INVALID_CONTROL_CHARS))) {
+    newText += text.substring(0, match.index);
+    newText += '&#x';
+    newText += text.charCodeAt(match.index).toString(16);
+    newText += ';';
+    text = text.substr(match.index + 1);
+  }
```

Comment:
"Instead of [this] while loop, could you just use the String's built-in [replace](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) method as follows, or is this less readable?"

```
return text.replace(RegExp(Blockly.utils.xml.INVALID_CONTROL_CHARS, "g"), function(match) {
  return '&#x' + match.charCodeAt(0).toString(16) + ';'
});
```

Reply:
That looks sooo much more readable. I'll switch it to that on Monday :D Thank you for the recommendation!